beater/handlers.go

```diff
@@ -87,11 +88,37 @@ func newMuxer(config *Config, report reporter) *http.ServeMux {
 	return mux
 }
 
+func concurrencyLimitHandler(config *Config, h http.Handler) http.Handler {
+	semaphore := make(chan struct{}, config.ConcurrentRequests)
+
+	release := func() {
+		<-semaphore
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		select {
+		case semaphore <- struct{}{}:
+			defer release()
+			h.ServeHTTP(w, r)
+		case <-time.After(config.MaxRequestQueueTime):
+			logger, ok := r.Context().Value(reqLoggerContextKey).(*logp.Logger)
+			if ok {
+				logger.Error("request rejected due to too many concurrent requests")
+			} else {
+				logp.NewLogger("request").Errorf("request rejected due to too many concurrent requests")
+			}
+			w.WriteHeader(503)
```

Comment:
I think `http.Error(w, http.StatusText(http.StatusServiceUnavailable), http.StatusServiceUnavailable)` is more idiomatic